lib/rbac/filterer.rb

```diff
@@ -291,7 +292,48 @@ def is_sti?(klass)
       klass.respond_to?(:finder_needs_type_condition?) ? klass.finder_needs_type_condition? : false
     end
 
-    def include_references(scope, klass, include_for_find, exp_includes)
+    # This is a very primitive way of determining whether we want to skip
+    # adding references to the query.
+    #
+    # For now, basically it checks if the caller has not provided :extra_cols,
+    # or if the MiqExpression can't apply the limit in SQL.  If both of those
+    # are true, then we don't add `.references` to the scope.
+    #
+    # If still invalid, there is an EXPLAIN check in #include_references that
+    # will make sure the query is valid and if not, will include the references
+    # as done previously.
+    def skip_references?(options, attrs)
+      options[:extra_cols].blank? && !attrs[:apply_limit_in_sql]
+    end
+
+    def include_references(scope, klass, include_for_find, exp_includes, skip)
+      if skip
+        # If we are in a transaction, we don't want to polute that
+        # transaction with a failed EXPLAIN.  We use a SQL SAVEPOINT (which is
+        # created via `transaction(:requires_new => true)`) to prevent that
+        # from being an issue (happens in tests with transactional fixtures)
+        #
+        # See https://stackoverflow.com/a/31146267/3574689
+        valid_skip = MiqDatabase.transaction(:requires_new => true) do
+          begin
+            ActiveRecord::Base.connection.explain(scope.to_sql)
+          rescue ActiveRecord::StatementInvalid => e
+            unless Rails.env.production?
+              warn "There was an issue with the Rbac filter without references!"
+              warn "Consider trying to fix this edge case in Rbac::Filterer!  Error Below:"
+              warn e.message
+              warn e.backtrace
+            end
+            # returns nil
+            raise ActiveRecord::Rollback
+          end
+        end
+        # If the result of the transaction is non-nil, then the block was
```

Comment:
@NickLaMuro when do we expect we'd try to skip references with an invalid scope?  What benefit does the above diagnostics provide over trying to use an invalid scope directly and blowing up?  In other words, why do you think the above is needed?  I'm asking because I don't know.

Reply:
> @NickLaMuro when do we expect we'd try to skip references with an invalid scope?

That is the "Million Dollar Question"...

I think this comment, and the comment from Keenan that spawned it kind sums it up:

https://github.com/ManageIQ/manageiq/pull/17141#issuecomment-374369307

But basically, I don't think there is a way to tell (properly) that we have all of our basis covered by removing `.references`.  This is basically an attempt to to make sure we aren't making any new breakages with this new fix.  That said, this is straight up a HACK, no question, but I don't know of a better way to handle this.  Before that, it was just the first two commits in this PR, and the `EXPLAIN` came in after the fact when I was questioning the work I did would be valid in all cases.  I don't necessarily think our tests cover all of our bases either, so that is why I am using this to attempt to make it so we don't make things worse with this "fix".

> What benefit does the above diagnostics provide over trying to use an invalid scope directly and blowing up?

Based on what I said above, the point of the diagnostics is to eventually get to a state where the `EXPLAIN` is deemed not needed.  The hope is that it would bombard the logs with output that would incentivize the developer to fix this in a more meaningful way instead of letting error continue.  Though, that said, I doubt these errors would show up in a QE environment, so maybe finding a way to trigger it in those environments might not be a bad idea either.

> In other words, why do you think the above is needed?

I realize that this was kind of answered in the above, but to re-state that answer:  I don't know.

I have added tests that simulate ways it could happen with the interfaces that `Rbac` provides, so it is "theoretically" possible, but whether this would actually be something that is triggered in practice is honestly an unknown.  I basically was trying to code defensively knowing that I don't know every possible report or automation script that might be going through this that could potentially break without the `.references` addition in place.

Reply:
> The change to the CloudTenancyMixin was made so that it can be in charge of it's own specific references instead of relying on a quirk of ActiveRecord, in which references(nil) will call references for all of the includes values.

I guess I was trying to understand what failing caller code would look like since I don't follow when this optimization would fail and inform the developer to fix it.

As an example, if you rolled back the changes to explicitly add `.references` to these models instead of relying on the quirk, would this explain debugging have caught an invalid scope?

Reply:
> As an example, if you rolled back the changes to explicitly add .references to these models instead of relying on the quirk, would this explain debugging have caught an invalid scope?

Yes, but I caught this myself when the tests failed, and this was prior to me adding the `EXPLAIN` concept after Keenan's first review.  Prior to that, I only had a single commit in this PR:

"Skip query references in Rbac when not needed"

I have rebased so many times at this point that the only thing that has stayed consistent is the commit title.

That said, the `CloudTenancyMixin` happened to be a prime example where we would want to fix things if this did get triggered, since it is a known and consistent scope.  Not sure that will always be the case with this for reporting though.

Reply:
> I guess I was trying to understand what failing caller code would look like since I don't follow when this optimization would fail and inform the developer to fix it.

You will get a Postrgresql error that the table is not part of the `FROM` (or `JOIN`).

Reply:
For a quick `rails console` example, you can run the following and see what happens when `.references` isn't present:

```ruby
irb> VmOrTemplate.includes(:host).where('"hosts"."name" = ?', 'foo')
#=> # BOOM!
irb> VmOrTemplate.includes(:host).references(:host).where('"hosts"."name" = ?', 'foo')
#=> # works!
```